src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -113,127 +135,378 @@ internal abstract class AbstractAddParameterCodeFixProvider<
 
-                        constructorsAndArgumentToAdd.Add(
-                            (constructor, argumentToAdd, arguments.IndexOf(argumentToAdd)));
+            // To keep the list of offered fixes short we create one menu entry per overload only
+            // as long as there are two or less overloads present. If there are more overloads we
+            // create two menu entries. One entry for non-cascading fixes and one with cascading fixes.
+            var fixes = codeFixData.Length <= 2
+                ? NestByOverload()
+                : NestByCascading();
+
+            context.RegisterFixes(fixes, context.Diagnostics);
+            return;
+
+            ImmutableArray<CodeAction> NestByOverload()
+            {
+                var builder = ArrayBuilder<CodeAction>.GetInstance(codeFixData.Length);
+                foreach (var data in codeFixData)
+                {
+                    // We create the mandatory data.CreateChangedSolutionNonCascading fix first.
+                    var title = GetCodeFixTitle(FeaturesResources.Add_parameter_to_0, data.Method, includeParameters: true);
+                    CodeAction codeAction = new MyCodeAction(
+                        title: title,
+                        data.CreateChangedSolutionNonCascading);
+                    if (data.CreateChangedSolutionCascading != null)
+                    {
+                        // We have two fixes to offer. We nest the two fixes in an inlinable CodeAction 
+                        // so the IDE is free to either show both at once or to create a sub-menu.
+                        var titleForNesting = GetCodeFixTitle(FeaturesResources.Add_parameter_to_0, data.Method, includeParameters: true);
+                        var titleCascading = GetCodeFixTitle(FeaturesResources.Add_parameter_to_0_and_overrides_implementations, data.Method, 
+                                                             includeParameters: true);
+                        codeAction = new CodeAction.CodeActionWithNestedActions(
+                            title: titleForNesting,
+                            ImmutableArray.Create(
+                                codeAction,
+                                new MyCodeAction(
+                                    title: titleCascading,
+                                    data.CreateChangedSolutionCascading)),
+                            isInlinable: true);
                     }
+
+                    // codeAction is now either a single fix or two fixes wrapped in a CodeActionWithNestedActions
+                    builder.Add(codeAction);
                 }
+
+                return builder.ToImmutableAndFree();
             }
 
+            ImmutableArray<CodeAction> NestByCascading()
+            {
+                var builder = ArrayBuilder<CodeAction>.GetInstance(2);
+
+                var nonCascadingActions = ImmutableArray.CreateRange<CodeFixData, CodeAction>(codeFixData, data =>
```

Comment:
so I guess assumption is all fix has non cascading fix? but only some has cascading fix?

Reply:
Exactly. Some method declarations can't *cascade* (e.g. static or private methods).

Reply:
tagging @jinujoseph @kuhlenh since it is design question whether we need non cascading fix when cascading fix exist.